cluster/mesos/cluster.go

```diff
@@ -0,0 +1,485 @@
+package mesos
+
+import (
+	"crypto/tls"
+	"errors"
+	"fmt"
+	"io"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/docker/docker/pkg/stringid"
+	"github.com/docker/swarm/cluster"
+	"github.com/docker/swarm/cluster/mesos/queue"
+	"github.com/docker/swarm/scheduler"
+	"github.com/docker/swarm/scheduler/node"
+	"github.com/docker/swarm/scheduler/strategy"
+	"github.com/docker/swarm/state"
+	"github.com/mesos/mesos-go/mesosproto"
+	mesosscheduler "github.com/mesos/mesos-go/scheduler"
+	"github.com/samalba/dockerclient"
+)
+
+// Cluster struct for mesos
+type Cluster struct {
+	sync.RWMutex
+
+	driver *mesosscheduler.MesosSchedulerDriver
+
+	eventHandler cluster.EventHandler
+	slaves       map[string]*slave
+	scheduler    *scheduler.Scheduler
+	options      *cluster.DriverOpts
+	store        *state.Store
+	TLSConfig    *tls.Config
+	master       string
+	pendingTasks *queue.Queue
+	offerTimeout time.Duration
+}
+
+var (
+	frameworkName    = "swarm"
+	dockerDaemonPort = "2375"
+	errNotSupported  = errors.New("not supported with mesos")
+)
+
+// NewCluster for mesos Cluster creation
+func NewCluster(scheduler *scheduler.Scheduler, store *state.Store, TLSConfig *tls.Config, master string, options cluster.DriverOpts) (cluster.Cluster, error) {
+	log.WithFields(log.Fields{"name": "mesos"}).Debug("Initializing cluster")
+
+	cluster := &Cluster{
+		slaves:       make(map[string]*slave),
+		scheduler:    scheduler,
+		options:      &options,
+		store:        store,
+		master:       master,
+		TLSConfig:    TLSConfig,
+		offerTimeout: 10 * time.Minute,
+	}
+
+	cluster.pendingTasks = queue.NewQueue()
+
+	// Empty string is accepted by the scheduler.
+	user, _ := options.String("mesos.user", "SWARM_MESOS_USER")
+
+	driverConfig := mesosscheduler.DriverConfig{
+		Scheduler: cluster,
+		Framework: &mesosproto.FrameworkInfo{Name: &frameworkName, User: &user},
+		Master:    cluster.master,
+	}
+
+	// Changing port for https
+	if cluster.TLSConfig != nil {
+		dockerDaemonPort = "2376"
```

Comment:
Perhaps we should have two consts in the file: `defaultDockerEnginePort` and `defaultDockerEngineTLSPort` and have a member in the struct named `dockerEnginePort`.

This check right here should decide whether to set `cluster.dockerEnginePort = defaultDockerEnginePort` or `cluster.dockerEnginePort = defaultDockerEngineTLSPort`.

Otherwise, as is, this is a global variable and we try to stay clear from those